coffee/rewriter.py

```diff
@@ -442,40 +543,55 @@ def preevaluate(self):
             self.header.children.remove(hoisted_loop)
         return self
 
-    def SGrewrite(self):
+    def sharing_graph_rewrite(self):
         """Rewrite the expression based on its sharing graph. Details in the
         paper:
 
-            On Optimality of Finite Element Integration, Luporini et. al.
+            An algorithm for the optimization of finite element integration loops
+            (Luporini et. al.)
         """
-        lda = loops_analysis(self.expr_info.linear_loops[0], key='symbol', value='dim')
-        sg_visitor = SharingGraph(self.expr_info, lda)
-
-        # Maximize the visibility of linear symbols
-        sgraph, mapper = sg_visitor.visit(self.stmt.rvalue)
-        if 'topsum' in mapper:
-            self.expand(mode='linear', subexprs=[mapper['topsum']])
-            sgraph, mapper = sg_visitor.visit(self.stmt.rvalue)
+        linear_dims = self.expr_info.linear_dims
+        other_dims = self.expr_info.out_linear_dims
+
+        # Maximize visibility of linear symbols
+        self.expand(mode='all')
+
+        # Make sure that potential reductions are not hidden away
+        lda = loops_analysis(self.header, value='dim')
+        self.reassociate(lambda i: (not lda[i]) + lda[i].issubset(set(other_dims)))
+
+        # Construct the sharing graph
+        nodes, edges = [], []
+        for i in summands(self.stmt.rvalue):
+            symbols = zip(*explore_operator(i))[0] if not isinstance(i, Symbol) else [i]
```

Comment:
If you merge with master, you get Python 3 style `zip` from `six.moves`, which means this doesn't work. You need to write `list(zip(...))[0]` instead of `zip(...)[0]`.